### ISSUE TITLE
CMake MSVC x64 Windows compilation failed

### ISSUE THREAD

**Issue by GitSparTV** [jtv/libpqxx#558]
Greetings. Trying to compile this library on windows.
My project cmake:
```cmake
cmake_minimum_required (VERSION 3.6)
set(CMAKE_CXX_STANDARD 23)

project(toolgun)

add_executable(toolgun main.cpp)

set(SKIP_BUILD_TEST ON)
set(BUILD_SHARED_LIBS ON)

set(PostgreSQL_ROOT "C:/Program Files/PostgreSQL/14/")

add_subdirectory(deps/libpqxx)
```

Here is how configuring went:
```
C:\Users\mails\Documents\C++\Toolgun\build>cmake .. -A x64
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19044.
-- The C compiler identification is MSVC 19.31.31105.0
-- The CXX compiler identification is MSVC 19.31.31105.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.31.31103/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.31.31103/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PostgreSQL: C:/Program Files/PostgreSQL/14/lib/libpq.lib (found version "14.2")
-- Looking for poll
-- Looking for poll - not found
-- Looking for PQencryptPasswordConn
-- Looking for PQencryptPasswordConn - not found
-- Looking for PQenterPipelineMode
-- Looking for PQenterPipelineMode - not found
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Generating config.h
-- Generating config.h - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/mails/Documents/C++/Toolgun/build
```

Here is how build went:
```
C:\Users\mails\Documents\C++\Toolgun\build>cmake --build . --config Debug
Microsoft (R) Build Engine version 17.1.0+ae57d105c for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/Users/mails/Documents/C++/Toolgun/deps/libpqxx/src/CMakeLists.txt
  array.cxx
  binarystring.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  blob.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  connection.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  cursor.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  encodings.cxx
  errorhandler.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  except.cxx
  field.cxx
  largeobject.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  notification.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  params.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  pipeline.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  result.cxx
  robusttransaction.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  row.cxx
  sql_cursor.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  strconv.cxx
  stream_from.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  stream_to.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  Generating Code...
  Compiling...
  subtransaction.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  time.cxx
  transaction.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  transaction_base.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\src\transaction_base.cxx(217,23): error C2672: 'binary_cast': no matching overloaded function found [C:
\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\src\transaction_base.cxx(217,38): error C7602: 'pqxx::binary_cast': the associated constraints are not
satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  util.cxx
  version.cxx
  wait.cxx
  Generating Code...
  Building Custom Rule C:/Users/mails/Documents/C++/Toolgun/CMakeLists.txt
  main.cpp
C:\Users\mails\Documents\C++\Toolgun\main.cpp(2,10): fatal error C1083: Cannot open include file: 'pqxx/pqxx': No such file or directory [C:\Users\mails\
Documents\C++\Toolgun\build\toolgun.vcxproj]
```

Basically:
- No 'pqxx/pqxx'
- binary_cast issue

**Comment by jtv** [comment]
I wish it could be more specific about the constraint issue.  Will have to
look in half an hour or so.

Shouldn't you tell CMake that your project depends on libpqxx?  The only
reference I see is the add_subdirectory, but that's not going to add
libpqxx's include directory to your include path, is it?

**Comment by GitSparTV** [comment]
That's what you have in [BUILDING-cmake.md](https://github.com/jtv/libpqxx/blob/master/BUILDING-cmake.md)
I did add target_link_libraries it didn't fix the issue

**Comment by GitSparTV** [comment]
main.cpp btw:
```c++
#include <iostream>
#include <pqxx/pqxx>

int main() {
    try {
        // Connect to the database.
        pqxx::connection C;
        std::cout << "Connected to " << C.dbname() << '\n';

        // Start a transaction.
        pqxx::work W{C};

        // Perform a query and retrieve all results.
        pqxx::result R{W.exec("SELECT name FROM employee")};

        // Iterate over results.
        std::cout << "Found " << R.size() << "employees:\n";
        for (auto row : R)
            std::cout << row[0].c_str() << '\n';

        // Perform a query and check that it returns no result.
        std::cout << "Doubling all employees' salaries...\n";
        W.exec0("UPDATE employee SET salary = salary*2");

        // Commit the transaction.
        std::cout << "Making changes definite: ";
        W.commit();
        std::cout << "OK.\n";
    } catch (std::exception const& e) {
        std::cerr << e.what() << '\n';
        return 1;
    }
    return 0;
}
```

**Comment by jtv** [comment]
Ah right.  Problem is I don't do much with CMake myself, so I'm entirely
dependent on contributions there.  Give me that half hour or so and I'll be
near a computer.

**Comment by jtv** [comment]
OK, have a look at #522.  It may have the CMake incantation you need.

**Comment by GitSparTV** [comment]
I'm testing on aarch64 (raspbian x64) my current cmake works, had to change 23 to 20.

**Comment by GitSparTV** [comment]
```cmake
cmake_minimum_required(VERSION 3.6)

IF (WIN32)
	set(CMAKE_CXX_STANDARD 23)
ELSE()
	set(CMAKE_CXX_STANDARD 20)
ENDIF()

project(toolgun)

add_executable(toolgun main.cpp)

set(SKIP_BUILD_TEST ON)
set(BUILD_SHARED_LIBS ON)

IF (WIN32)
	find_package(PostgreSQL PATHS "C:/Program Files/PostgreSQL/14/")
ENDIF()

find_package(PostgreSQL REQUIRED)

add_subdirectory(deps/libpqxx)
target_link_libraries(toolgun PUBLIC pqxx PostgreSQL::PostgreSQL)
```

To work on both systems, still binary_cast issues

**Comment by tt4g** [comment]
It could be a bug in the MSVC compiler.
The Compiler Explorer can build the following minimal code.

@GitSparTV Could you try to see if you can build it in your Visual Studio IDE?

Compiler Explorer: https://godbolt.org/z/xo3d8bPaT

```cpp
#include <string>
#include <string_view>
#include <type_traits>

template<typename TYPE>
using strip_t = std::remove_cv_t<std::remove_reference_t<TYPE>>;

template<typename TYPE>
std::basic_string_view<std::byte> binary_cast(TYPE const &data)
{
  return {
    reinterpret_cast<std::byte const *>(
      const_cast<strip_t<decltype(*std::data(data))> const *>(
        std::data(data))),
    std::size(data)};
}

template<typename CHAR, typename SIZE>
std::basic_string_view<std::byte> binary_cast(CHAR const *data, SIZE size)
{
  static_assert(sizeof(CHAR) == 1);
  return {
    reinterpret_cast<std::byte const *>(data),
    check_cast<std::size_t>(size, "binary data size")};
}

void main(void)
{
    std::string_view str = "Hello, World!";

    binary_cast(str);
}
```

**Comment by GitSparTV** [comment]
@tt4g 
With c++latest it compiles fine, here are the warnings:
```
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xmemory(154,5): warning C4365: 'argument': conversion from 'long' to 'unsigned int', signed/unsigned mismatch
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.31.31103\include\xmemory(164,5): warning C4365: 'argument': conversion from 'long' to 'unsigned int', signed/unsigned mismatch
1>C:\Users\mails\Documents\C++\test\Source.cpp(25,17): warning C4326: return type of 'main' should be 'int' instead of 'void'
```

On C++17 it errors.

**Comment by GitSparTV** [comment]
Btw, little bit off-topic, but why strip_t is used when C++11 has std::decay?

**Comment by jtv** [comment]
Those warnings with c++latest are all outside of libpqxx — two in the compiler's headers, and one in the application code.  (Yes, `main()` should return `int`, not `void`.)

As for `std::decay` — it does a lot more than what I wanted.  As the documentation says, I intend to replace `pqxx::strip_t` with `std::remove_cvref`.

**Comment by tt4g** [comment]
Many CMake features will not work unless defined after `project`. This is probably the case with `CMAKE_CXX_STANDARD` as well.
Try the following changes:

```diff
 cmake_minimum_required(VERSION 3.6)
-
-IF (WIN32)
-	set(CMAKE_CXX_STANDARD 23)
-ELSE()
-	set(CMAKE_CXX_STANDARD 20)
-ENDIF()

 project(toolgun)
+
+IF (WIN32)
+	set(CMAKE_CXX_STANDARD 23)
+ELSE()
+	set(CMAKE_CXX_STANDARD 20)
+ENDIF()

 add_executable(toolgun main.cpp)
```

**Comment by GitSparTV** [comment]
Still errors on binary_cast

**Comment by tt4g** [comment]
It may be due to the CMake cache. have you deleted the CMake cache?

**Comment by GitSparTV** [comment]
I erase build folder every time I change something

**Comment by GitSparTV** [comment]
Current status:
Cmake:
```cmake
cmake_minimum_required(VERSION 3.6)

project(toolgun)

IF (WIN32)
	set(CMAKE_CXX_STANDARD 23)
ELSE()
	set(CMAKE_CXX_STANDARD 20)
ENDIF()

add_executable(toolgun main.cpp)

IF (WIN32)
	IF (POLICY CMP0074)
		cmake_policy(SET CMP0074 NEW)
	ENDIF()

	# find_package(PostgreSQL PATHS "C:/Program Files/PostgreSQL/14/")
	set(PostgreSQL_ROOT "C:/Program Files/PostgreSQL/14/")
ENDIF()

find_package(PostgreSQL REQUIRED)

set(SKIP_BUILD_TEST ON)
set(BUILD_SHARED_LIBS ON)

add_subdirectory(deps/libpqxx)

target_link_libraries(toolgun PUBLIC pqxx PostgreSQL::PostgreSQL)
```

Configure:
```
C:\Users\mails\Documents\C++\Toolgun\build>cmake .. -A x64
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19044.
-- The C compiler identification is MSVC 19.31.31105.0
-- The CXX compiler identification is MSVC 19.31.31105.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.31.31103/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.31.31103/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PostgreSQL: C:/Program Files/PostgreSQL/14/lib/libpq.lib (found version "14.2")
-- Looking for poll
-- Looking for poll - not found
-- Looking for PQencryptPasswordConn
-- Looking for PQencryptPasswordConn - not found
-- Looking for PQenterPipelineMode
-- Looking for PQenterPipelineMode - not found
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Generating config.h
-- Generating config.h - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/mails/Documents/C++/Toolgun/build
```

Build:
```
C:\Users\mails\Documents\C++\Toolgun\build>cmake --build . --config Debug
Microsoft (R) Build Engine version 17.1.0+ae57d105c for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/Users/mails/Documents/C++/Toolgun/deps/libpqxx/src/CMakeLists.txt
  array.cxx
  binarystring.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  blob.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  connection.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  cursor.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  encodings.cxx
  errorhandler.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  except.cxx
  field.cxx
  largeobject.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  notification.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  params.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  pipeline.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  result.cxx
  robusttransaction.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  row.cxx
  sql_cursor.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  strconv.cxx
  stream_from.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  stream_to.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  Generating Code...
  Compiling...
  subtransaction.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  time.cxx
  transaction.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  transaction_base.cxx
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,23): error C2672: 'binary_cast': no matching overloaded function foun
d [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/binarystring.hxx(220,1): error C7602: 'pqxx::binary_cast': the associated constraints are
not satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\src\transaction_base.cxx(217,23): error C2672: 'binary_cast': no matching overloaded function found [C:
\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\src\transaction_base.cxx(217,38): error C7602: 'pqxx::binary_cast': the associated constraints are not
satisfied [C:\Users\mails\Documents\C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
C:\Users\mails\Documents\C++\Toolgun\deps\libpqxx\include\pqxx/util.hxx(296): message : see declaration of 'pqxx::binary_cast' [C:\Users\mails\Documents\
C++\Toolgun\build\deps\libpqxx\src\pqxx.vcxproj]
  util.cxx
  version.cxx
  wait.cxx
  Generating Code...
```

**Comment by tt4g** [comment]
It appears that `pqxx::binrary_cast` does not work well with MSVC.
The following code builds with gcc, but fails with MSVC.

Compiler Explorer: https://godbolt.org/z/6qE7h4dxs

```cpp
#include <iterator>
#include <ranges>
#include <utility>
#include <string_view>

// From: libpqxx/include/pqxx/types.hxx

template<typename TYPE>
using strip_t = std::remove_cv_t<std::remove_reference_t<TYPE>>;

template<std::ranges::range CONTAINER>
using value_type = decltype(*std::begin(std::declval<CONTAINER>()));

template<typename DATA>
concept potential_binary = std::ranges::contiguous_range<DATA> and
  (sizeof(value_type<DATA>) == 1);

// From: libpqxx/include/pqxx/util.hxx

template<potential_binary TYPE>
std::basic_string_view<std::byte> binary_cast(TYPE const &data)
{
    static_assert(sizeof(value_type<TYPE>) == 1);
  return {
    reinterpret_cast<std::byte const *>(
      const_cast<strip_t<decltype(*std::data(data))> const *>(
        std::data(data))),
    std::size(data)};
}


int main(void)
{
    std::string_view foo = "foo";

    binary_cast(foo);
}
```

I found the cause, but did not have time to look for a solution.
I will look again for a solution when I have more time.

**Comment by GitSparTV** [comment]
Tested on my machine, same behaviour

**Comment by jtv** [comment]
I do have automatic test builds on appveyor, so perhaps I can upgrade the
compiler and/or the C++ version to reproduce the problem.

**Comment by GitSparTV** [comment]
It works if you replace
`using value_type = decltype(*std::begin(std::declval<CONTAINER>()));`
to
`using value_type = strip_t<decltype(*std::begin(std::declval<CONTAINER>()))>;`

Because `decltype(*std::begin(std::declval<CONTAINER>()));` returns `const char&`, strip_t transforms back to `char` which works for msvc. But that's certainly a bug because it's `const char&` in gcc but it works

**Comment by jtv** [comment]
(I've reproduced the problem in appveyor, by setting the compiler to compile C++20 or C++23.  The default (presumably C++17 on Visual Studio 2022) did work.)

I think you're right that `value_type` should be stripped of cv-qualifiers and references.  Bear with me while I wait for #559 to go through tests.

**Comment by jtv** [comment]
Thanks @GitSparTV and @tt4g.

Does the final paragraph in [BUILDING-cmake.md](../blob/master/BUILDING-cmake.md) need updating?

**Comment by GitSparTV** [comment]
I can finish the template I have right now 

I guess I will also switch to vcpkg because I see I'm missing libssl and other libraries. They are not in pq folder

**Comment by jtv** [comment]
I would expect libpq's CMake setup to declare those dependencies as needed.  The only _direct_ dependency libpqxx itself has is on libpq.

**Comment by GitSparTV** [comment]
Or we can fix it so everything works without manual fixing by others individually.

Libpqxx doesn't tell about .dll you have to ship with the program.

**Comment by jtv** [comment]
It would be nice to have it all work out of the box, but I think it would be unwise to duplicate the knowledge of what DLLs libpq needs.  That list can change from one libpq version to another.

**Comment by jtv** [comment]
Can anyone think of a good way to incorporate instructions for install, that does not rely too much on a specific libpq version, the user's platform, or the individual choices of shared vs. static libraries?

**Comment by GitSparTV** [comment]
Here is my latest working cmake, cross-platform for windows and raspberry pi (probably for any linux)

```cmake
cmake_minimum_required(VERSION 3.6)

project(toolgun CXX)

IF (WIN32)
	# Ranges are available only in /std:c++latest (aka 23) on msvc
	set(CMAKE_CXX_STANDARD 23)
ELSE()
	set(CMAKE_CXX_STANDARD 20)
ENDIF()

add_executable(toolgun main.cpp)

# If the library cannot be found, provide the root directly
# Example: -DPostgreSQL_ROOT="C:/Program Files/PostgreSQL/14/"
IF (PostgreSQL_ROOT AND POLICY CMP0074)
	cmake_policy(SET CMP0074 NEW)
ENDIF()

find_package(PostgreSQL REQUIRED)

set(SKIP_BUILD_TEST ON)

IF (WIN32)
	set(BUILD_SHARED_LIBS ON)
ELSE()
	set(BUILD_SHARED_LIBS OFF)
ENDIF()

add_subdirectory(deps/libpqxx)
target_link_libraries(toolgun PUBLIC pqxx PostgreSQL::PostgreSQL)
```

With this cmake it configured, builds and compiles without issues.
Issues comes from dynamic linking:
First dependency: `pqxx.dll` which is in `*build directory of the project*/deps/libpqxx/src/Debug/`
After that: `LIBPQ.dll`. Located in `*postgresql location*/lib/`
After that: `libcrypto-1_1-x64.dll`, `libssl-1_1-x64.dll` and `libintl-9.dll`. Located in `*postgresql location*/bin/`
After that: `libwinpthread-1.dll`, `libiconv-2.dll`. Located in the same folder.
And only after that I get successful run: `connection to server at "localhost" (::1), port 5432 failed: fe_sendauth: no password supplied`. I either didn't setup the postgres or didn't provide password, but it executes.

**Comment by jtv** [comment]
Thanks, I'm trying to update the documentation based on this.  Still puzzled though as to why CMake does not take care of the libraries that the dependencies require.

I did notice however a repeated typo in `BUILDING-cmake.md`: it said "PostgresSQL" when of course it should say "PostgreSQL".

Oh, I got that warning about `<ranges>` and it linked to an MSVC bug which IIRC said you could test for `IF (MSVC)` to determine whether you need to go straight to C++23.

**Comment by tt4g** [comment]
Unfortunately, CMake does not have accurate information about libraries built by tools other than CMake.
Since PostgreSQL is installed in almost all locations, it only scans candidate installation directories to find libraries. Therefore, it may not be able to detect dependent libraries.
Windows, in particular, does not have as organized rules as UNIX Like OS for library installation locations, so dependent libraries may not be discovered.
This problem is more pronounced for build tools that are additionally installed in Windows, such as MinGW.

The names of missing libraries are reported, so we often check the error message and add the missing library.

**Comment by GitSparTV** [comment]
Ok, so should we leave this to the developer of the program to add them manually?
Maybe at least pqxx.dll could be moved to program folder

**Comment by tt4g** [comment]
Yes, libraries that CMake could not found should be made to be found by the developer. However, there is no need to change the installation path of the library.

For PostgreSQL, there is an option that allows CMake to give installation path hints.
Set the CMake variable `PostgreSQL_ROOT` to the PostgreSQL installation base directory (this is an option you are already using).

If the system library cannot be found, it can be added to the link target by adding the system library to the CMake variable `CMAKE_CXX_IMPLICIT_LINK_LIBRARIES` (https://cmake.org/cmake/help/v3.23/variable/CMAKE_LANG_IMPLICIT_LINK_LIBRARIES.html#variable:CMAKE_%3CLANG%3E_IMPLICIT_LINK_LIBRARIES).

**Comment by GitSparTV** [comment]
What about libpqxx which cmake builds for my project?

**Comment by tt4g** [comment]
Since you are doing a cross-platform build with CMake, add a link target to CMake according to the target host.

Example:

```cmake
IF (WIN32)
		set(BUILD_SHARED_LIBS  ON)
        # Add linker search path.
        list(APPEND CMAKE_CXX_IMPLICIT_LINK_DIRECTORIES "*postgresql location*/bin/")
        set(toolgun_ADDITIONAL_LIBRARIES libcrypto-1_1-x64 libssl-1_1-x64 libintl-9 libwinpthread-1 libiconv-2)
ELSE()
		set(BUILD_SHARED_LIBS  OFF)
        set(toolgun_ADDITIONAL_LIBRARIES)
ENDIF()

add_subdirectory(deps/libpqxx)
target_link_libraries(toolgun PUBLIC pqxx PostgreSQL::PostgreSQL ${toolgun_ADDITIONAL_LIBRARIES})
````

Or add additional library paths to `target_link_libraries` with full paths.

Example:

```cmake
IF (WIN32)
		set(BUILD_SHARED_LIBS  ON)
ELSE()
		set(BUILD_SHARED_LIBS  OFF)
ENDIF()

add_subdirectory(deps/libpqxx)
target_link_libraries(toolgun PUBLIC pqxx PostgreSQL::PostgreSQL
    "*postgresql location*/bin/libcrypto-1_1-x64.dll"
    "*postgresql location*/bin/libssl-1_1-x64.dll"
    "*postgresql location*/bin/libintl-9.dll"
    "*postgresql location*/bin/libwinpthread-1.dll"
    "*postgresql location*/bin/libiconv-2.dll"
)
````

See also https://cmake.org/cmake/help/v3.23/command/target_link_libraries.html

**Comment by GitSparTV** [comment]
I don't see how I can copy pqxx.dll with CMake. Cmake builds it, so it knows everything about it

**Comment by jtv** [comment]
Doesn't CMake also know how to _install_ it?

**Comment by jtv** [comment]
(I realise you don't want your end-users to run CMake, but sometimes you can gather all the libraries you need by doing a fake install to a working directory.)

**Comment by jtv** [comment]
I'm updating the documentation based on this discussion, but obviously I'm out of my depth when it comes to CMake.  Perhaps you could have a look at the changes in #561 and comment?

**Comment by tt4g** [comment]
> I don't see how I can copy pqxx.dll with CMake. Cmake builds it, so it knows everything about it

@GitSparTV You can use Cmake generator-expressions and `add_custom_command()`.

- https://cmake.org/cmake/help/v3.23/manual/cmake-generator-expressions.7.html#target-dependent-queries
- https://cmake.org/cmake/help/v3.23/command/add_custom_command.html

By adding the following `add_custom_command()`, you should be able to copy the pqxx libraries to `<OUTPUT_PATH>` (replace with the destination directory) after the `toolgun` build is complete:

```cmake
add_custom_command(TARGET toolgun POST_BUILD
	COMMAND ${CMAKE_COMMAND} -E $<TARGET_FILE:pqxx> <OUTPUT_PATH>
	VERBATIM
	COMMAND_EXPAND_LISTS
)
```

`$<TARGET_FILE:pqxx>` is a CMake generator-expression that is converted to the full path to the pqxx artifact.

**Comment by GitSparTV** [comment]
Alright, one not, is it fine that libpqxx has potentially conflicting variables like `SKIP_BUILD_TEST`?

**Comment by tt4g** [comment]
No conflict.

The pqxx test is defined as a CMake target `runner`. When `SKIP_BUILD_TEST=ON`, `runner` is just not defined (https://github.com/jtv/libpqxx/blob/7.7.2/test/CMakeLists.txt#L9).

**Comment by GitSparTV** [comment]
What i mean by that is what if I have other library which has this variable. I expect to see `LIBPQXX_SKIP_BUILD_TEST`

**Comment by tt4g** [comment]
Are you worried that `SKIP_BUILD_TEST` will be overwritten when you import other libraries with `add_subdirectory()`?
Then don't worry, CMake has different variable scopes for the child project imported by `add_subdirectory()` and for the parent project.
Overwriting a variable in the child project does not effect the variable in the parent project.

Variables in the parent project can be rewritten if the `PARENT_SCOPE` option is specified in `set()`. However, this option is usually not used.

https://cmake.org/cmake/help/v3.23/command/set.html

**Comment by GitSparTV** [comment]
Thanks, I think we figured out everything. Would be great if people can just copy cmake example in building-cmake to start quickly

**Comment by jtv** [comment]
By the way, thanks for this folks.  I've been distracted but I'll update
that PR over the coming days, so people will have a starting point.  And
I'll ask for your feedback.
